internal/graph/check.go

```diff
@@ -141,313 +139,226 @@ type CheckHandlerFunc func(ctx context.Context) (*ResolveCheckResponse, error)
+	if finalErr != nil {
+		return nil, finalErr
 	}
 
-	resp = &ResolveCheckResponse{
-		Allowed: true,
-	}
-
-	return
+	// If the loop completes without any "false" outcomes or errors, the result is "true".
+	return finalResult, nil
 }
 
 // exclusion implements a CheckFuncReducer that requires a 'base' CheckHandlerFunc to resolve to an allowed
-// outcome and a 'sub' CheckHandlerFunc to resolve to a falsey outcome. The base and sub computations are
-// handled concurrently relative to one another.
-func exclusion(ctx context.Context, concurrencyLimit int, handlers ...CheckHandlerFunc) (*ResolveCheckResponse, error) {
+// outcome and a 'sub' CheckHandlerFunc to resolve to a falsey outcome.
+func exclusion(ctx context.Context, _ int, handlers ...CheckHandlerFunc) (*ResolveCheckResponse, error) {
 	if len(handlers) != 2 {
 		return nil, fmt.Errorf("%w, expected two rewrite operands for exclusion operator, but got '%d'", openfgaErrors.ErrUnknown, len(handlers))
 	}
 
-	span := trace.SpanFromContext(ctx)
-
-	limiter := make(chan struct{}, concurrencyLimit)
-
 	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
 	baseChan := make(chan checkOutcome, 1)
 	subChan := make(chan checkOutcome, 1)
 
-	var wg sync.WaitGroup
-
-	defer func() {
-		cancel()
-		wg.Wait()
-		close(baseChan)
-		close(subChan)
-	}()
-
-	baseHandler := handlers[0]
-	subHandler := handlers[1]
-
-	limiter <- struct{}{}
-	wg.Add(1)
 	go func() {
-		recoveredError := panics.Try(func() {
-			defer func() {
-				wg.Done()
-				<-limiter
-			}()
-
-			resp, err := baseHandler(ctx)
-			baseChan <- checkOutcome{resp, err}
-		})
-
-		if recoveredError != nil {
-			baseChan <- checkOutcome{nil, fmt.Errorf("%w: %s", ErrPanic, recoveredError.AsError())}
-		}
+		concurrency.TrySendThroughChannel(ctx, runHandler(ctx, handlers[0]), baseChan)
+		close(baseChan)
 	}()
-
-	limiter <- struct{}{}
-	wg.Add(1)
 	go func() {
-		recoveredError := panics.Try(func() {
-			defer func() {
-				wg.Done()
-				<-limiter
-			}()
-
-			resp, err := subHandler(ctx)
-			subChan <- checkOutcome{resp, err}
-		})
-		if recoveredError != nil {
-			subChan <- checkOutcome{nil, fmt.Errorf("%w: %s", ErrPanic, recoveredError.AsError())}
-		}
+		concurrency.TrySendThroughChannel(ctx, runHandler(ctx, handlers[1]), subChan)
+		close(subChan)
 	}()
 
-	response := &ResolveCheckResponse{
-		Allowed: false,
-	}
-
-	var baseErr error
-	var subErr error
+	var baseErr, subErr error
 
-	for i := 0; i < len(handlers); i++ {
+	// Loop until we have received one result from each of the two channels.
+	resultsReceived := 0
+	for resultsReceived < 2 {
 		select {
-		case baseResult := <-baseChan:
-			if baseResult.err != nil {
-				telemetry.TraceError(span, baseResult.err)
-				baseErr = baseResult.err
+		case <-ctx.Done():
+			return nil, ctx.Err()
+
+		case res, ok := <-baseChan:
+			if !ok {
+				baseChan = nil // Stop selecting this case.
 				continue
 			}
+			resultsReceived++
```

Comment:
Question - why do we want to increment this after the `if !ok` in line 317?  What will happen in the case where the channel is closed immediately (for some reason without ever receiving a valid message)?  In  this case, would we be stuck in this loop until timeout?

Reply:
>  why do we want to increment this after the if !ok in line 317?

Because `!ok` means no message was received. Thus this loop is just to exit out.

> What will happen in the case where the channel is closed immediately (for some reason without ever receiving a valid message)? In this case, would we be stuck in this loop until timeout?

The only case in which that can happen if it's the `ctx` is cancelled. Which we have a handle for. Otherwise you receive a message.

There are units that actually cover these questions too.

Reply:
So, if I understand correctly, the intend if we see !ok without ever receiving a valid message is to return context timeout error in line 314.  Can we add comment to call this out?

Reply:
It doesn't need to be a timeout. Whichever cancelation reason we receive, we would exit through case of `ctx.Done`. 

Notice the producer of this channel will always attempt to write to the channel. The only reason were you would get no message, and a channel closed event (because given its a buffered channel you will receive everything in the order that it happens) is if the context is no longer active.

This is somewhat explicit here:
```
go func() {
		concurrency.TrySendThroughChannel(ctx, runHandler(ctx, handlers[0]), baseChan)
		close(baseChan)
	}()

```